### DISCUSSION TITLE
ImageMagick convert command takes 30 times longer to convert a pgm image to png when the image consists of horizontal lines compared to when it consists of vertical lines

### DISCUSSION THREAD

**Discussion by JohnB141592** [ImageMagick/ImageMagick#7524] [Help]
### ImageMagick version

7.0.10-61

### Operating system, version and so on

Cygwin under Windows 11

### Description

I have C++ code that writes images in pgm format. I then use ImageMagick's command line convert command to convert those pgm images into png format. The images are 5000 x 5000 pixels in size. I find that when the image consists of a lot of random vertical lines, it consistently takes about 2-3 seconds to convert. When the image consists of random horizontal lines, it consistently takes 45-90 seconds to convert. This is the case even when the images are otherwise identical, i.e. the horizontal one is just the vertical one rotated 90 degrees. This difference does not occur when converting to either jpg or gif, just png. Why is this happening and how can I get around it? Thanks!

**Comment by snibgo** [comment]
>  ImageMagick's command line convert command ...

You have v7. `convert` is now deprecated, and you should use `magick` instead.

Please show the command you used, and a sample file.

I guess the difference is in the PNG compression. These can be changed with `-quality` or `-define png:compression-filter=` etc. See https://imagemagick.org/script/defines.php

**Reply by JohnB141592** [reply]
Thanks for your response! I tried it with the "magick" command and got the same results. The command I used was as follows:

magick TestSample_Layer12.pgm TestSample_Layer12.png

I can't include the file "TestSample_Layer12.pgm" directly as a sample because this forum does not support that file type. So I gzipped it and attached that: [TestSample_Layer12.pgm.gz](https://github.com/user-attachments/files/16472963/TestSample_Layer12.pgm.gz) I tried values 0-4 for the "-define png:compression-filter=" and it made no difference. Is there some other value I should try? What values should I try for the "-quality" option?

**Comment by snibgo** [comment]
I can't reproduce the major difference in performance.
```
C:\web\im>timec %IMG7%magick TestSample_Layer12.pgm x1.png

TimeC C:\im\ImageMagick-7.1.1-20-Q16-HDRI\magick TestSample_Layer12.pgm x1.png
Seconds: 0.86


C:\web\im>timec %IMG7%magick TestSample_Layer12.pgm -rotate 90 x2.png

TimeC C:\im\ImageMagick-7.1.1-20-Q16-HDRI\magick TestSample_Layer12.pgm -rotate 90 x2.png
Seconds: 1.02
```

My IM uses libpng 1.6.40. What does your IM use?
```
magick -list format |grep -i png

     APNG  rw+   Animated Portable Network Graphics
             See http://www.libpng.org/pub/mng/ for details about the JNG
      MNG* rw+   Multiple-image Network Graphics (libpng 1.6.40)
             See http://www.libpng.org/pub/mng/ for details about the MNG
      PNG* rw-   Portable Network Graphics (libpng 1.6.40)
             See http://www.libpng.org/ for details about the PNG format.
    PNG00* rw-   PNG inheriting bit-depth, color-type from original, if possible
    PNG24* rw-   opaque or binary transparent 24-bit RGB
    PNG32* rw-   opaque or transparent 32-bit RGBA
    PNG48* rw-   opaque or binary transparent 48-bit RGB
    PNG64* rw-   opaque or transparent 64-bit RGBA
     PNG8* rw-   8-bit indexed with optional binary transparency
```

**Reply by JohnB141592** [reply]
`magick -list format | grep -i png

     APNG  rw+   Animated Portable Network Graphics
           See http://www.libpng.org/pub/mng/ for details about the JNG
      MNG* rw+   Multiple-image Network Graphics (libpng 1.6.39,1.6.42)
           See http://www.libpng.org/pub/mng/ for details about the MNG
      PNG* rw-   Portable Network Graphics (libpng 1.6.39,1.6.42)
           See http://www.libpng.org/ for details about the PNG format.
    PNG00* rw-   PNG inheriting bit-depth, color-type from original, if possible
    PNG24* rw-   opaque or binary transparent 24-bit RGB (zlib 1.2.13,1.3.1)
    PNG32* rw-   opaque or transparent 32-bit RGBA
    PNG48* rw-   opaque or binary transparent 48-bit RGB
    PNG64* rw-   opaque or transparent 64-bit RGBA
     PNG8* rw-   8-bit indexed with optional binary transparency`

**Comment by snibgo** [comment]
> (libpng 1.6.39,1.6.42)

You IM uses two versions of libpng? Seems strange. But I expect those versions are compatible.

I can't reproduce the problem, and don't know what else to suggest.

**Reply by JohnB141592** [reply]
OK, thanks for your help. I think I'll just switch to jpg instead of png.

**Comment by snibgo** [comment]
JPG uses lossy compression. If that is a problem, you might use TIFF, either with no compression, or with LZW or ZIP compression.

**Reply by JohnB141592** [reply]
Is there a way to coax IM into making a tiff with 32 bit depth instead of 8 bit depth?

**Comment by fmw42** [comment]
See the formats page at https://imagemagick.org/script/formats.php#google_vignette for tiff.  

`"To specify a single-precision floating-point format, use (-depth 32 and) -define quantum:format=floating-point. Set the depth to 64 for a double-precision floating-point format."`

**Reply by JohnB141592** [reply]
When I enter the command

` magick -depth 32 -define quantum:format=floating-point TestSample_Layer00.pgm TestSample_Layer00.tiff`

it says "aborted" and no output is produced. Is that the correct command?

**Comment by snibgo** [comment]
You need to specify `-depth` and `-define` for the output, but not the input. So put them before the output, but not before the input.
```
 magick TestSample_Layer00.pgm -depth 32 -define quantum:format=floating-point TestSample_Layer00.tiff
```

**Reply by JohnB141592** [reply]
It worked. Thanks!
